51: WIP handling knowledge-based quiz markdown.

glenpike opened the pull request:
Fixes #51

Create parser for handling `---question---` blocks in the markdown.
Question block has specific structure, so calls it's own methods.
Create spec's for the examples in the documentation.
Indentation is not very good, but not sure how to do that nicely without adding nokogiri 😟

glenpike committed:
51: WIP handling knowledge-based quiz markdown.

Basic parsing, not quite passing test.
- Code needs splitting up a bit
- Spacing / indentation is an issue.
- Adding extra stuff to the markup that _might_ be needed.
- Don't have 'project name' - how to get?

glenpike committed:
51: Question Meta

glenpike committed:
51: Changing round to work with simpler design

- Only handling separate 'question' pages now.
- Remove a load of methods.
- Split examples - example_question sort of working...
- TODO - get 'blocks' parsing behaving.

glenpike committed:
51: question_blocks_in_feedback parsing/rendering okay

glenpike committed:
51: Handling questions with one feedback item

glenpike committed:
51: Fix spacing / indentation of reference html and tweak indexing of feedback.

glenpike committed:
51: Namespace the KnowledgeQuiz stuff.

glenpike committed:
51: Fixing for example questions.

glenpike commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -172,6 +202,79 @@ def self.convert_task_to_html(task)
+          HEREDOC
+        end
+  
+        def self.convert_choices_to_html(text)
```
This is quite a chunky method, tried to figure ways to break down without passing objects around too much or allowing called methods to mutate things, but not much luck.
Also difficult to have a single regex to check for feedback and split a 'choice' into it's constituent parts.  Feel that this regex would possibly be quite complicated.

patch0 commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -112,6 +119,29 @@ def self.convert_print_only_to_html(content)
 
+      def self.convert_knowledge_quiz_question_to_html(question, _indent)
+        question_match = QUESTION_REGEXP.match(question)
+        return '' if question_match.nil? || question_match.length() < 2
```
Useless empty brackets after `length`
```suggestion
        return '' if question_match.nil? || question_match.length < 2
```

patch0 commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -1,8 +1,15 @@
     module Kramdown
       YAML_FRONT_MATTER_REGEXP = /\n\s*---\s*\n(.*?)---(.*)/m
-      RADIO_REGEXP = /\(\s?\)\s*(?<text>.*)/
+      RADIO_REGEXP = /\((?<check>[\sx*]{0,1})\)\s*(?<text>.*)/m
+      CHOICE_BLOCK_REGEXP = %r{^(?=#{::Kramdown::Parser::Kramdown::OPT_SPACE}- \([\sx*]?\)\s*.*)}m
+      FEEDBACK_REGEXP_PARTIAL = "#{::Kramdown::Parser::Kramdown::OPT_SPACE}---[ \t]*feedback[ \t]*---(.*?)---[ \t]*\/feedback[ \t]*---"
+      CHOICE_FEEDBACK_REGEXP = %r{#{FEEDBACK_REGEXP_PARTIAL}}m
+      SINGLE_FEEDBACK_REGEXP = %r{\A#{FEEDBACK_REGEXP_PARTIAL}}m
+      QUESTION_REGEXP = %r{(.*?)^#{::Kramdown::Parser::Kramdown::OPT_SPACE}---[ \t]*choices[ \t]*---(.*?)---[ \t]*\/choices[ \t]*---}m
```
These should all be frozen, as they're constants.

```suggestion
      RADIO_REGEXP = /\((?<check>[\sx*]{0,1})\)\s*(?<text>.*)/m.freeze
      CHOICE_BLOCK_REGEXP = %r{^(?=#{::Kramdown::Parser::Kramdown::OPT_SPACE}- \([\sx*]?\)\s*.*)}m.freeze
      FEEDBACK_REGEXP_PARTIAL = "#{::Kramdown::Parser::Kramdown::OPT_SPACE}---[ \t]*feedback[ \t]*---(.*?)---[ \t]*\/feedback[ \t]*---".freeze
      CHOICE_FEEDBACK_REGEXP = %r{#{FEEDBACK_REGEXP_PARTIAL}}m.freeze
      SINGLE_FEEDBACK_REGEXP = %r{\A#{FEEDBACK_REGEXP_PARTIAL}}m.freeze
      QUESTION_REGEXP = %r{(.*?)^#{::Kramdown::Parser::Kramdown::OPT_SPACE}---[ \t]*choices[ \t]*---(.*?)---[ \t]*\/choices[ \t]*---}m.freeze
```

patch0 commented on lib/kramdown_rpf/kramdown.rb:
```diff
@@ -199,6 +226,7 @@ class KramdownRPF < ::Kramdown::Parser::GFM
       NEW_PAGE_PATTERN   = %r{^#{OPT_SPACE}---[ \t]*new-page[ \t]*---}m
       NO_PRINT_PATTERN   = %r{^#{OPT_SPACE}---[ \t]*no-print[ \t]*---(.*?)---[ \t]*\/no-print[ \t]*---}m
       PRINT_ONLY_PATTERN = %r{^#{OPT_SPACE}---[ \t]*print-only[ \t]*---(.*?)---[ \t]*\/print-only[ \t]*---}m
+      KNOWLEDGE_QUIZ_QUESTION_PATTERN = %r{^#{OPT_SPACE}---[ \t]*question[ \t]*---(.*?)---[ \t]*\/question[ \t]*---}m
```
All these should be frozen too
```suggestion
      KNOWLEDGE_QUIZ_QUESTION_PATTERN = %r{^#{OPT_SPACE}---[ \t]*question[ \t]*---(.*?)---[ \t]*\/question[ \t]*---}m.freeze
```

patch0 commented on lib/kramdown_rpf/kramdown.rb:
```diff
@@ -164,6 +185,11 @@ def convert_print_only(el, _indent)
       end
 
+      # Convert :quiz -> LaTEX (not implemented)
+      def convert_quiz(_el, _opts)
```
If you don't care about the options, you can just use `*`:

```suggestion
      def convert_quiz(*)
```

patch0 commented on lib/kramdown_rpf/kramdown.rb:
```diff
@@ -149,6 +165,11 @@ def convert_hints(_el, _opts)
       end
 
+      # Convert :convert_knowledge_quiz_question -> LaTEX (not implemented)
+      def convert_knowledge_quiz_question(_el, _opts)
```
```suggestion
      def convert_knowledge_quiz_question(*)
```

patch0 commented on lib/kramdown_rpf/kramdown.rb:
```diff
@@ -112,6 +123,11 @@ def convert_print_only(el, _indent)
       end
 
+      # Convert :quiz -> Markdown (not implemented)
+      def convert_quiz(_el, _opts)
```
```suggestion
      def convert_quiz(*)
```

patch0 commented on lib/kramdown_rpf/kramdown.rb:
```diff
@@ -97,6 +103,11 @@ def convert_hints(_el, _opts)
       end
 
+      # Convert :convert_knowledge_quiz_question -> Markdown (not implemented)
+      def convert_knowledge_quiz_question(_el, _opts)
```
If you don't care about the options, you can use `*` as a glob.
```suggestion
      def convert_knowledge_quiz_question(*)
```

patch0 commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -172,6 +202,79 @@ def self.convert_task_to_html(task)
+          choices.each.with_index do |choice, index|
+            choice_match = RADIO_REGEXP.match(choice)
+
+            next unless choice_match && choice_match['text']
```
You can use safe navigation here (although it is a bit weird with the array operator `[]`)

```suggestion
            next unless choice_match&.[]('text')
```

patch0 commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -172,6 +202,79 @@ def self.convert_task_to_html(task)
+
+            choice_with_feedback = choice_match['text'].strip.split(CHOICE_FEEDBACK_REGEXP)
+
+            checked = choice_match['check'] == 'x' || choice_match['check'] == '*'
```
I think this might be a bit more idiomatic.
```suggestion
            checked = %w(x *).include?(choice_match['check'])
```

patch0 replied:
Or even better.. you could have a VALID_CHECK_MARKS array somewhere, and use that in the regex and here.  It saves an allocation of an array on each iteration.

patch0 commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -172,6 +202,79 @@ def self.convert_task_to_html(task)
+            checked = choice_match['check'] == 'x' || choice_match['check'] == '*'
+            choice_html += convert_label_to_html(choice_with_feedback[0], index, checked)
+
+            next if choice_with_feedback.length() < 2
```
```suggestion
            next if choice_with_feedback.length < 2
```

patch0 commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -172,6 +202,79 @@ def self.convert_task_to_html(task)
+          if feedback_html.size.positive?
+            feedback_html = <<~HEREDOC
+              <ul class="knowledge_quiz__feedback">
+                #{feedback_html.strip}
```
Should the `strip` be applied before `.size.positive?`?

patch0 reviewed (commented):
These are some clean-up changes that could be made.

Also the target ruby version in .rubocop.yml is set to 2.3, which should be upped to 2.7 I suspect.

Commit:
Update lib/kramdown_rpf/rpf.rb

Co-authored-by: Patrick Cherry <patch0@users.noreply.github.com>

Commit:
Apply suggestions from code review

Co-authored-by: Patrick Cherry <patch0@users.noreply.github.com>

glenpike committed:
51: Fixing versions of required gems

glenpike committed:
51: Make sure you spec development dependencies, just not versions.

glenpike committed:
51: Tidy-ups

glenpike committed:
51: Fix easy Rubocop errors

sHtev commented on .rubocop.yml:
```diff
@@ -1,5 +1,5 @@
 AllCops:
-  TargetRubyVersion: 2.3
+  TargetRubyVersion: 2.7
```
I'd be careful - as a library we should aim for wider support, but on the other hand only we use it, on the gripping hand, we are only on 2.5 on Lambda runtime

glenpike committed:
Target Ruby 2.5 for Rubocop - AWS lambda is running 2.5

sHtev commented on examples/knowledge_quiz/example_question.html:
```diff
@@ -0,0 +1,69 @@
+    <div class="knowledge_quiz__blurb">
+      <p>The bug sprite starts facing upwards.</p>
+
+<p><img src="./img/q3-1.png" alt="Bug facing upwards" /></p>
```
are we unable to get this to indent, not a big deal, but it makes me v. sad

glenpike replied:
Could not figure how to do quickly originally - was thinking of looking at it after I'd got everything working.

Guess I'll have to do a lot of this:
```
foo_html.join('      ').strip
```

glenpike replied:
I had a go at this - was doing okay until I hit the `<ul>` - but it leaves the code looking really messy / weird, so decided that it's better that this is legible / consistent.  If we want pretty-printed html, we should run it through nokogiri.

sHtev commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -101,7 +110,7 @@ def self.convert_hints_to_html(hints)
 
       def self.convert_new_page_to_html
-        ::Kramdown::Document.new("<div class=\"c-print-page-break\" />", KRAMDOWN_OPTIONS).to_html
+        ::Kramdown::Document.new('<div class="c-print-page-break" />', KRAMDOWN_OPTIONS).to_html
```
this is changed here (disagree using single quotes for output regardless of escaping)...

sHtev commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -112,20 +121,43 @@ def self.convert_print_only_to_html(content)
         ::Kramdown::Document.new("<div class=\"u-print-only\">\n#{content}</div>", KRAMDOWN_OPTIONS).to_html
```
...but this is not changed here

sHtev replied:
food for thought - https://anti-pattern.com/always-use-double-quoted-strings-in-ruby

glenpike replied:
That one interpolates a variable, so not changed.
Rubocop flags one, but not t'other

sHtev commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -112,20 +121,43 @@ def self.convert_print_only_to_html(content)
 
+      def self.convert_knowledge_quiz_question_to_html(question, _indent)
+        question_match = QUESTION_REGEXP.match(question)
+        return '' if question_match.nil? || question_match.length < 2
```
""?

glenpike replied:
See below

sHtev commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -112,20 +121,43 @@ def self.convert_print_only_to_html(content)
       def self.convert_quiz_to_html(quiz)
         content = YAML_FRONT_MATTER_REGEXP.match(quiz)
-        return '' if content.nil? || content.length() < 2
+        return '' if content.nil? || content.length < 2
```
""?

sHtev replied:
(I know I wrote this, just wondering about inconsistency)

sHtev commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -140,7 +172,7 @@ def self.convert_quiz_to_html(quiz)
 
               <div class="c-project-quiz__content">
-                #{radio_inputs.join("      ").strip}
+                #{radio_inputs.join('      ').strip}
```
rationale? is this a rubocop style thing - I bet @patch0 can point me to the style guide, I just have it ingrained to use double quotes for output strings

sHtev commented on lib/kramdown_rpf/rpf.rb:
```diff
@@ -172,6 +204,81 @@ def self.convert_task_to_html(task)
+            choices.shift
+          end
+
+          choices.each.with_index do |choice, index|
```
still think this could be a separate method that just deals with a single choice

sHtev replied:
that is calling one in this loop, or regexing a containing `--- choice ---` or something

glenpike replied:
Should I pass in variables to mutate or something else? (It appends to `choice_html` and `feedback_html` - that's the bit I couldn't get around)

patch0 commented on lib/kramdown_rpf/kramdown.rb:
```diff
@@ -6,202 +8,230 @@ module Converter
+        RPF::Plugin::Kramdown.convert_task_to_html(element.value)
       end
     end
 
```
I'm not clear why we have all these methods defined in the next two classes, but they could be shortened to:

```ruby
    # Stupid module name
    #
    # This saves us defining all the methods in Html again for Kramdown and Latex classes and have them 
    # raise a NotImplementedError in each case.
    module RaiseNotImplementedForUndefinedConvertMethods
      def method_missing(method_name, *args, &block)
        raise NotImplementedError if method_name.to_s.start_with?('convert_')

        super
      end

      def respond_to_missing?(method_name, include_private = false)
        method_name.to_s.start_with?('convert_') || super
      end
    end

    class Kramdown
      include RaiseNotImplementedForUndefinedConvertMethods
    end

    class Latex
      include RaiseNotImplementedForUndefinedConvertMethods
    end
```

patch0 replied:
Frankly I'd be happy to leave them as NoMethodError, but heh.

patch0 commented on lib/kramdown_rpf/kramdown.rb:
```diff
@@ -6,202 +8,230 @@ module Converter
+        raise NotImplementedError
+      end
+
+      # Convert :quiz -> LaTEX (not implemented)
```
At least spell LaTeX correctly ;)

glenpike replied:
Okay, if you do too

```
# This saves us defining all the methods in Html again for Kramdown and Latex classes and have them 
    # raise a NotImplementedError in each case.
```

patch0 replied:
That's the class name :p

glenpike committed:
51: Review tweaks

- consistent match variable name
- Meta programming fix for non-implemented methods.

glenpike committed:
51: Update version to 0.11
